android/src/main/java/com/vydia/RNUploader/UploadProgress.kt

```diff
@@ -0,0 +1,62 @@
+package com.vydia.RNUploader
+
+import android.content.Context
+import android.os.Handler
+import android.os.Looper
+import androidx.work.WorkManager
+import com.vydia.RNUploader.UploaderModule.Companion.WORKER_TAG
+
+// Stores and aggregates total progress from all workers
+class UploadProgress {
+
+  companion object {
+    private fun storage(context: Context) =
+      context.getSharedPreferences("RNFileUpload-Progress", Context.MODE_PRIVATE)
+
+    fun set(context: Context, uploadId: String, bytesUploaded: Long, fileSize: Long) =
+      storage(context).edit()
+        .putLong("$uploadId-uploaded", bytesUploaded)
+        .putLong("$uploadId-size", fileSize)
+        .commit()
+
+    fun remove(context: Context, uploadId: String) =
+      storage(context).edit()
+        .remove("$uploadId-uploaded")
+        .remove("$uploadId-size")
+        .commit()
+
+    fun total(context: Context): Double {
+      val storage = storage(context)
+
+      val totalBytesUploaded = storage.all.keys
+        .filter { it.endsWith("-uploaded") }
+        .sumOf { storage.getLong(it, 0L) }
+
+      val totalFileSize = storage.all.keys
+        .filter { it.endsWith("-size") }
+        .sumOf { storage.getLong(it, 0L) }
+
+      if (totalFileSize == 0L) return 0.0
+      return (totalBytesUploaded.toDouble() * 100 / totalFileSize)
+    }
+
+    private val handler = Handler(Looper.getMainLooper())
+
+    // Attempt to clear in 2 seconds. This straightforward approach
+    // enables the final worker to reset the overall progress.
+    // Clearing progress ensures the notification starts at 0% next time.
+    fun scheduleClearing(context: Context) =
+      handler.postDelayed({ clearIfNeeded(context) }, 2000)
+
+    fun clearIfNeeded(context: Context) {
+      val workManager = WorkManager.getInstance(context)
+      val works = workManager.getWorkInfosByTag(WORKER_TAG).get()
+      if (works.any { !it.state.isFinished }) return
+
+      val storage = storage(context)
+      val editor = storage.edit()
+      storage.all.keys.forEach { key -> editor.remove(key) }
+      editor.commit()
```

Comment:
Android Studio gave me a warning on this line that we could use `editor.apply()` to do this asynchronously, not sure why it didn't warn in the other 2 places.

Reply:
I deliberately use `commit` in these places. `apply` is async and will send the `commit` to a different thread, which makes sense if we don't want to block the current thread, but these `commits` mostly run on coroutines so we'd want them to be synchronous for the results to be accurate. The only case it runs on the main thread is at start up, which is ok.